### ISSUE TITLE
Roundabout instruction improvement

### ISSUE THREAD

**Issue by jansoe** [graphhopper/graphhopper#319]
As a follow up to the implementation of roundabout instructions (#316) some improvements are yet to be implemented:

+ [ ] improved detection of direction of rotation by not relying on turn geometry, but purely on the edges' order. For a clean implementation there should be no CH shortcuts in the graph (#116)
+ [ ] a more accurate assesment of the transit angle through the roundabout in case of drive-up ways. To this end either one calculates entrance/exit direction not only from the last two waypoints but interpolates across the last X meters, or one avarages waypoints of both the incoming and outgoing direction of the edge.
+ [x] more fine grained text instructions and rephrase "take exit X in direction dummy street" to "take exit X onto dummy street"
+ [ ] exit numbers are counted as exits the vehicle (i.e. encoder) could take. This might lead to confusion as e.g for bikes this could increase the exit number (https://graphhopper.com/maps/?point=52.513955%2C13.344462&point=52.511598%2C13.357036&vehicle=bike&elevation=true&layer=TF%20Cycle) but a user might assume that only paved streets are counted. Vice versa this could reduce the number of counted exits for trucks as some exits might be prohibited for them.
+ [ ] There is a counting problem if the incoming edge its last node is identical to the first node of the next outgoing edge. See the screenshot of miniui (right side), which is taken from old data in the file core/files/monaco.osm.gz [here](https://graphhopper.com/maps/?point=43.735782%2C7.416962&point=43.735992%2C7.416909)

![bhaaabjj](https://cloud.githubusercontent.com/assets/129644/6354300/c2504fde-bc4e-11e4-81f8-45e1e0d793de.png)

**Comment by devemux86** [comment]
I propose to make ```public``` the [RoundaboutInstruction.getExitNr](https://github.com/graphhopper/graphhopper/blob/master/core/src/main/java/com/graphhopper/util/RoundaboutInstruction.java#L76) to be able to handle more flexible the ```RoundaboutInstruction``` (like the others) being able to get its elements.

**Comment by karussell** [comment]
(If so, please rename to getExitNumber :))

**Comment by devemux86** [comment]
It would be also helpful if the ```RoundaboutInstruction```'s ```finished``` and ```continuedStreet``` variables get public get methods, since they participate in instruction building.

**Comment by jansoe** [comment]
please note that ```isContinuedStreet```  is already public.
I will make  ```getExitNr``` public and rename it to ```getExitNumber```.
  
I hesitate to make ```finished``` public as it's main purpose is internal consitency control, but you are right it takes part in the instruction building if the goal is within a roundabout.  With the upcoming improved roundabout instruction in mind I would introduce an indicator variable  that indicates if a roundabout either has to be entered and left | only entered | only left.

**Comment by devemux86** [comment]
I know you have mentioned that the roundabout instructions are to be improved.
Because right now do we really need the instruction with [0 exit](https://github.com/graphhopper/graphhopper/blob/master/core/src/main/java/com/graphhopper/util/RoundaboutInstruction.java#L137)?
Isn't confusing for the user?

**Comment by jansoe** [comment]
Yes I think it might be confusing. But I also think it is a very unusal case that a route is to end within a roundabout and therefore this instruction is temporarly acceptable. 
I totally agree that the refined instructions should be available as soon as possible, but I also wanted to give the community the chance to give feedback.

**Comment by mattfbacon** [comment]
We just encountered the exit counting issue with bicycles. Here was the discussion on OSM (In german): <https://www.openstreetmap.org/note/4470785#map=19/49.902639/10.067153&layers=N>. Has anything been done to address this? I don't understand why it would be different for bicycles because they also have to follow traffic rules like one-way streets.
